### DISCUSSION TITLE
ceph connection

### DISCUSSION THREAD

**Discussion by QZB349549482** [ceph/go-ceph#881] [Q&A] [answered]
I need to read the ceph configuration file and then link ceph. Currently I know only two methods, ReadDefaultConfigFile() and ReadConfigFile(). Is there any other way to read.
```
// ReadDefaultConfigFile configures the connection using a Ceph configuration
// file located at default locations.
func (c *Conn) ReadDefaultConfigFile() error {
	ret := C.rados_conf_read_file(c.cluster, nil)
	return getError(ret)
}

// ReadConfigFile configures the connection using a Ceph configuration file.
func (c *Conn) ReadConfigFile(path string) error {
	cPath := C.CString(path)
	defer C.free(unsafe.Pointer(cPath))
	ret := C.rados_conf_read_file(c.cluster, cPath)
	return getError(ret)
}
```
I wonder if the following code would help
```
// ParseConfigArgv configures the connection using a unix style command line
// argument vector.
//
// Implements:
//
//	int rados_conf_parse_argv(rados_t cluster, int argc,
//	                          const char **argv);
func (c *Conn) ParseConfigArgv(argv []string) error {
	if c.cluster == nil {
		return ErrNotConnected
	}
	if len(argv) == 0 {
		return ErrEmptyArgument
	}
	cargv := make([]*C.char, len(argv))
	for i := range argv {
		cargv[i] = C.CString(argv[i])
		defer C.free(unsafe.Pointer(cargv[i]))
	}

	ret := C.rados_conf_parse_argv(c.cluster, C.int(len(cargv)), &cargv[0])
	return getError(ret)
}

// ParseCmdLineArgs configures the connection from command line arguments.
//
// This function passes a placeholder value to Ceph as argv[0], see
// ParseConfigArgv for a version of this function that allows the caller to
// specify argv[0].
func (c *Conn) ParseCmdLineArgs(args []string) error {
	argv := make([]string, len(args)+1)
	// Ceph expects a proper argv array as the actual contents with the
	// first element containing the executable name
	argv[0] = argvPlaceholder
	for i := range args {
		argv[i+1] = args[i]
	}
	return c.ParseConfigArgv(argv)
}
```

**Comment by phlogistonjohn** [comment, marked as answer]
> I need to read the ceph configuration file and then link ceph. Currently I know only two methods, ReadDefaultConfigFile() and ReadConfigFile(). Is there any other way to read.

> [... snip ...]

> I wonder if the following code would help


Hi there,
Unfortunately your statements appear somewhat contradictory to me. You start off by asking for methods to read the ceph configuration file and then imply the two functions to perform that action are somehow not suitable for that purpose.  Then you ask if the functions for reading ceph command line arguments are suitable.

So, it's not clear to me what you are really trying to do.  The library provides three ways to configure the underlying ceph libraries. The functions for reading configuration files (ReadDefaultConfigFile, ReadConfigFile as you know), then the functions for parsing command line options (ParseConfigArgv, ParseCmdLineArgs which you asked about) and then there's [SetConfigOption](https://pkg.go.dev/github.com/ceph/go-ceph@v0.21.0/rados#Conn.SetConfigOption) which is the most simple and direct way to set an *individual* configuration option via code.

If you need more details than I've shared here you'll have to elaborate more on what you are trying to accomplish. Thanks!

**Reply by QZB349549482** [reply]
Thanks for answer! I want to read ceph config file  . But I do not know there is another way to do it . (ps:ReadDefaultConfigFile, ReadConfigFile is not suitable for me )

**Reply by phlogistonjohn** [reply]
> Thanks for answer! I want to read ceph config file . But I do not know there is another way to do it . 

Not that I know of

> (ps:ReadDefaultConfigFile, ReadConfigFile is not suitable for me )

Why?

**Reply by QZB349549482** [reply]
I am sorry for it. The method ReadConfigFile will work . Thank you very much

**Reply by phlogistonjohn** [reply]
Great to hear, thanks for letting me know!
